mypy_django_plugin/transformers/models.py

```diff
@@ -587,6 +599,133 @@ def run_with_model_cls(self, model_cls: Type[Model]) -> None:
             self.add_new_node_to_model_class("_meta", Instance(options_info, [Instance(self.model_classdef.info, [])]))
 
 
+class MetaclassAdjustments(ModelClassInitializer):
+    @classmethod
+    def adjust_model_class(cls, ctx: ClassDefContext) -> None:
+        """
+        For the sake of type checkers other than mypy, some attributes that are
+        dynamically added by Django's model metaclass has been annotated on
+        `django.db.models.base.Model`. We remove those attributes and will handle them
+        through the plugin.
+        """
+        if ctx.cls.fullname != fullnames.MODEL_CLASS_FULLNAME:
+            return
+
+        does_not_exist = ctx.cls.info.names.get("DoesNotExist")
+        if does_not_exist is not None and isinstance(does_not_exist.node, Var) and not does_not_exist.plugin_generated:
+            del ctx.cls.info.names["DoesNotExist"]
+
+        multiple_objects_returned = ctx.cls.info.names.get("MultipleObjectsReturned")
+        if (
+            multiple_objects_returned is not None
+            and isinstance(multiple_objects_returned.node, Var)
+            and not multiple_objects_returned.plugin_generated
+        ):
+            del ctx.cls.info.names["MultipleObjectsReturned"]
+
+        return
+
+    def get_exception_bases(self, name: str) -> List[Instance]:
+        bases = []
+        for model_base in self.model_classdef.info.direct_base_classes():
+            exception_base_sym = model_base.names.get(name)
+            if (
+                # Base class is a Model
+                model_base.metaclass_type is not None
+                and model_base.metaclass_type.type.fullname == fullnames.MODEL_METACLASS_FULLNAME
+                # But base class is not 'models.Model'
+                and model_base.fullname != fullnames.MODEL_CLASS_FULLNAME
+                # Base class also has a generated exception base e.g. 'DoesNotExist'
+                and exception_base_sym is not None
+                and exception_base_sym.plugin_generated
+                and isinstance(exception_base_sym.node, TypeInfo)
+            ):
+                bases.append(Instance(exception_base_sym.node, []))
+
+        return bases
+
+    @cached_property
+    def is_model_abstract(self) -> bool:
+        meta = self.model_classdef.info.names.get("Meta")
+        # Check if 'abstract' is declared in this model's 'class Meta' as
+        # 'abstract = True' won't be inherited from a parent model.
+        if meta is not None and isinstance(meta.node, TypeInfo) and "abstract" in meta.node.names:
+            for stmt in meta.node.defn.defs.body:
+                if (
+                    # abstract =
+                    isinstance(stmt, AssignmentStmt)
+                    and len(stmt.lvalues) == 1
+                    and isinstance(stmt.lvalues[0], NameExpr)
+                    and stmt.lvalues[0].name == "abstract"
+                    and (
+                        (
```

Comment:
Let's break this down into separate conditions to have more readability.

1. Ending with `and stmt.lvalues[0].name == "abstract"`
2. About `abstract = True`
3. About `abstract: Literal[True]`